### PR TITLE
Multithreading Support

### PR DESCRIPTION
**Pull request by johnzl-777**
Enable multithreading support for emulation via the ODE solver.

## Design

For each matrix type the Hamiltonian object holds (SparseMatrixCSC, Permutation Matrix, Diagonal), a parallelized version of `mul!` using `Polyester` exists that will take advantage of all the threads the user starts Julia with (e.g. `julia -t num_of_threads`).

The user can select between three possible backends for this parallelization:
* `BloqadeExpr` - The default, single-threaded implementation. Even with multiple threads, this just calls the standard `mul!` from `SparseArrays`/`LuxurySparse`.
* `ThreadedSparseCSR` - converts all matrices to CSR format and performs standard Sparse Matrix-Dense Vector multiplication. The `bmul!` function was originally part of the `ThreadedSparseCSR` library but was copied into `BloqadeExpr` due to its restriction to an older version of `Polyester` that could have issues with threads counts > 64. 
* `ParallelMergeCSR` - uses QuEra's `ParallelMergeCSR.jl` library which has better performance on larger systems/matrices with rows that have a large difference in the number of nonzero elements (e.g. Blockade Subspace aware Hamiltonian matrices).

## Usage

1. Import `BloqadeExpr`.
2. Check the existing backend through the global variable `BloqadeExpr.backend`. By default it is set to "BloqadeExpr". 
3. To change to a multithreaded implementation, call `BloqadeExpr.set_backend()` and pass in a string (should be "ParallelMergeCSR" or "ThreadedSparseCSR")
4. Exit Julia, then restart with the desired number of threads through `julia -t num_threads`
5. Upon importing Bloqade/BloqadeExpr again the proper backend should be set

~~Users set a `BLOQADE_THREADS` environment variable to the desirable number of threads to use as well as toggling the Sparse Matrix-Dense Vector backend through `BloqadeExpr.set_backend` (unexported, and works via Preferences.jl).~~

## Benchmarks

Benchmarks were conducted on an Intel Core i7-12700 with 16 GB of RAM using Ubuntu 20.04.4 LTS.

The `DP8()` solver was selected for all benchmarks (should be the new default Bloqade solver since #532 ), with the geometry being a Chain Lattice that had constant Rabi Frequency of 4 * 2\pi rad/us applied to it with a distance of 5.74 um between atoms and evolution emulated for 1.6 microseconds.

BenchmarkTools.jl was used to run each problem 200 times and the median taken to produce the values/graphs below. 

image

image

This table shows the median run time for a fixed system of 16 atoms in the same configuration above, but with the number of threads gradually increased using the `ThreadedSparseCSR` backend:

| # of threads | Median Run Time (s) (out of 200 runs) |
| --- | --- |
| 1 | 10.10 |
| 2 | 7.09 | 
| 4 | 4.98 |
| 8 | 3.88 |

### CODE REVIEW

**Review by Roger-luo — COMMENTED** [review]
Can we attach a (simple one) benchmark on how each backend performs with multi thread enabled?

**Comment by johnzl-777** [comment]
> Can we attach a (simple one) benchmark on how each backend performs with multi thread enabled?

Can do! Currently fixing some compatibility issues that are popping up 😅

**Comment by weinbe58** [comment]
John has already run a few benchmarks already!

**Comment by Roger-luo** [comment]
> John has already run a few benchmarks already!

would be nice if we can attach that in this PR

File: lib/BloqadeExpr/src/linalg.jl [hunk 1]
```diff
@@ -5,3 +5,66 @@ function LinearAlgebra.mul!(C::AbstractVecOrMat, A::StepHamiltonian, B::Abstract
     end
     return C
 end
+
+# if BloqadeExpr was the backend of choice, then ThreadedMatrix will have the SparseMatrixCSC type
+# in which case we just dispatch to standard mul!
+LinearAlgebra.mul!(C, A::ThreadedMatrix{<:SparseMatrixCSC}, B, α, β) = SparseArrays.mul!(C, A.matrix, B, α, β)
```
Comment (Roger-luo) [inline lib/BloqadeExpr/src/linalg.jl:11]: I think we need some basica tests for these

**Review by Roger-luo — COMMENTED** [review]
the code looks good to me, but we should have some basic tests that covers the missing lines reported by codecov

File: lib/BloqadeExpr/src/linalg.jl (hunk 1 above)
Comment (johnzl-777) [reply lib/BloqadeExpr/src/linalg.jl:11]: Can do!

**Review by Roger-luo — APPROVED** [review]

File: lib/BloqadeExpr/Project.toml [hunk 2]
```diff
@@ -11,7 +11,11 @@ LaTeXStrings = "b964fa9f-0449-5b57-a5c2-d3ea65f4040f"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 LuxurySparse = "d05aeea4-b7d4-55ac-b691-9e7fabb07ba2"
 MLStyle = "d8e11817-5142-5d16-987a-aa16d5891078"
+ParallelMergeCSR = "c4b929cc-6141-4ba4-8849-280a863346e1"
```
Comment (Roger-luo) [inline lib/BloqadeExpr/Project.toml:14]: not right now but we should consider moving these to an extension to reduce the package size thus the latency when we swap out BloqadeExpr with BloqadePulse

**Pull request merged**